### PR TITLE
Allow uploading reports to iotile.cloud

### PR DESCRIPTION
**Pull request by timburke**
## Major Enhancements

- Adds support for uploading reports to iotile.cloud
- Fixes CloudSession to reuse thread pools rather than create 10 new threads every time you create a new CloudSession (which happens somewhat often under the hood because it's assumed to be cheap)
- Adds `ActionReport` template that has a table of links with tooltips and short descriptions.  This should be used for `LiveReport` objects that are meant to provide a list of possible next steps to take after reading the report.  The action table is created automatically based on what links a subclass has added with `add_action`  calls.
- Add `prepare_render()` hook in `LiveReport` to alllow us to move initialization code out of the `__init__.py` method of all subclasses.  ActionReport uses this to relax when users can call `add_action` by moving the actual jinja2 templating of the action table to `prepare_render()`.
- Fixes CSS issues on IE 11 since IE apparently does not autoscale svg images correctly so you need to pass explicit heights and widths.

### Action Tables

Action tables are an optional convenience feature that you can use to create a table of links at the top of a report.  Each link has 4 fields and is created by calling `add_action` after subclassing from `ActionReport`:

- link_text: The actual text of the link
- link_target: What is used as the href for the link (can be a javascript function)
- short_desc: A line line description that is shown to the right of the link in the action table
- help_text: An arbitrary HTML blob that is shown in a tooltip if specified to allow for longer explanations of what an action does without making the table too wordy.  If help_text is passes, a `?` appears next to the link that shows the tool tip.

### Action Table (showing tooltip)

image

### Action Table (without tooltip)

image

### CODE REVIEW

File: iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            'acl': 'public-read' if public is True else 'private',
+            'content_type': self._get_mime_type(file_name)
+        }
+
+        resource = self._api.report.generated(report_id).uploadurl
+
+        return resource.url(), payload
+
+    @classmethod
+    def _get_mime_type(cls, filename):
```
Comment (dkarchmer) [inline iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py]: Just FYI: The cloud is now doing this exact same thing if you don't pass the content_type

File: iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                raise ArgumentError("The group provided did not have org or source slug information in its source_info", slug=slug, org=org)
+        elif org is None or slug is None:
+            raise UsageError("If you do not specify an AnalysisGroup you must pass an org and a slug to this function", org=org, slug=slug)
+
+        report_id = self._create_report(label, slug, org)
+        self._logger.debug("Created report id: %s", report_id)
+
+        keys = self._clean_file_paths(files)
+
+        urls, fields = self._get_signed_urls(report_id, keys)
```
Comment (dkarchmer) [inline iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py]: It seems you first get all signed URLs, and then upload them.
Any reason why you don't do it as you go? i.e. get url and upload.
The reason I am asking is that all the signed URLs have an expiration time. I don't think this is an issue for this application but in theory there could be a situation where the URL expires before the upload happens

File: iotile_analytics_interactive/iotile_analytics/interactive/scripts/analytics_host.py [hunk 3]
```diff
@@ -113,6 +114,10 @@ def build_args():
     parser.add_argument('-c', '--no-confirm', action="store_true", help="Do not confirm the analysis that you are about to perforn and prompt for parameters")
     parser.add_argument('-l', '--list', action="store_true", help="List all known analysis types without running one")
     parser.add_argument('-b', '--bundle', action="store_true", help="Bundle the rendered output into a zip file")
+    parser.add_argument('-w', '--web-push', action="store_true", help="Push the resulting report to iotile.cloud.")
+    parser.add_argument('--web-push-label', type=str, default=None, help="Set the label used when pushing a report to iotile.cloud")
+    parser.add_argument('--web-push-org', type=str, default=None, help="Override the org given in the analysisgroup and force it to be this")
```
Comment (dkarchmer) [inline iotile_analytics_interactive/iotile_analytics/interactive/scripts/analytics_host.py]: The org should not be needed, as it can be inferred by the device/block.
I assume we don't want to allow uploading to a different org

File: iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py (hunk 2 above)
Comment (timburke) [reply iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py]: Interesting.  We can refactor this if that becomes an issue.  The reason is that the parallelization for the requests is handled by `CloudSession` and there wasn't a good way to express 'post this' then get the result, call this function on it and then post again, whereas I could just make a `post_multiple` function that did it in stages but I see the point.

Right now getting the signatures is fast, just a few seconds for 300 of them.

File: iotile_analytics_interactive/iotile_analytics/interactive/scripts/analytics_host.py (hunk 3 above)
Comment (timburke) [reply iotile_analytics_interactive/iotile_analytics/interactive/scripts/analytics_host.py]: Good point.  Let me refactor this.
